[CALCITE-5160] ANY/SOME, ALL operators should support collection expressions

dssysolyatin opened the pull request.

bchapuis commented on core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:
```diff
@@ -4049,6 +4081,99 @@ private static class HopImplementor implements TableFunctionCallImplementor {
+        return NULL_EXPR;
+      }
+
+      // final T _quantify_left_value = <>
```
To me, it is not obvious why the following lines are commented.

dssysolyatin replied:
I have added more comments to explain why this code has been commented.

bchapuis commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -3636,6 +3636,44 @@ public static Object[] array(Object... args) {
+  /**
+   * Returns whether there is an element in {@code list} for which {@code predicate} is true.
+   * Also, if {@code predicate} returns null for any element of {@code list}
+   * and there are no true comparison result is obtained, the result will be null, not false.
```
This sentence should probably be reworked (i.e. "there are" and "is obtained").

dssysolyatin replied:
Fixed.

bchapuis commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -3636,6 +3636,44 @@ public static Object[] array(Object... args) {
+  /**
+   * Returns whether {@code predicate} is true for all elements of {@code list}.
+   * Also, if {@code predicate} returns null for any element of {@code list}
+   * and there are no false comparison result is obtained, the result will be null, not true.
```
This sentence should probably be reworked (i.e. "there are" and "is obtained").

dssysolyatin replied:
Fixed.

bchapuis commented on core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:
```diff
@@ -5451,7 +5461,11 @@ ImmutableList<RelNode> retrieveCursors() {
+        if (subQuery == null && (kind == SqlKind.SOME || kind == SqlKind.ALL)) {
+          break;
+        }
+        assert subQuery != null;
```
This assert can probably be removed as requireNonNull is called right after.

dssysolyatin replied:
Probably it can be removed. However, I feel that it's important to keep it because requireNonNull is applied to `subQuery.expr`. If a NullPointerException is thrown, it could be difficult to identify whether the root cause lies with `subQuery` or `subQuery.expr`.

zoudan replied:
Agree with @bchapuis, it is meaningless to call 'assert' here.

dssysolyatin replied:
@zoudan @bchapuis  I tend to disagree with you guys. How to distinguish NPE in these pieces of code without debugging ?
```
subQuery = null;
// .... Some code
requireNonNull(subQuery.expr);
```

```
subQuery = new SubQuery(node, logic, clause);
subQuery.expr = null;
// .... Some code
requireNonNull(subQuery.expr);
```

Both of these snippets result in an NPE with an identical stacktrace. The only way to understand what the root cause of the NPE (e.g `subQuery` or `subQuery.expr`) is debugging this code. However, getting additional information (e.g., queries) for debugging can be hard in a production environment due to security concerns. So, It's really important to have as much information as possible to understand and fix the problem.

bchapuis commented on testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:
```diff
@@ -130,27 +133,22 @@
 /**
  * Contains unit tests for all operators. Each of the methods is named after an
  * operator.
- *
```
This file has probably been reformatted by the IDE.

dssysolyatin replied:
Fixed.

bchapuis reviewed (commented):
I was interested in seeing how these operators were implemented and added a couple of comments. As I'm not yet fully familiar with this area of the code, you should not rely on my review to merge these changes.

zoudan commented on core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:
```diff
@@ -4049,6 +4081,101 @@ private static class HopImplementor implements TableFunctionCallImplementor {
+            // the result of ALL will be null, not true.
+            : BuiltInMethod.COLLECTION_NULLABLE_ALL.method;
+      default:
+        throw new IllegalArgumentException("Unknown quantify operator" + kind
```
add space after `operator `

zoudan commented on core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:
```diff
@@ -4049,6 +4081,101 @@ private static class HopImplementor implements TableFunctionCallImplementor {
+      }
+    }
+  }
+
```
remove one empty line

dssysolyatin commented:
@bchapuis @zoudan I fixed the address reviews except https://github.com/apache/calcite/pull/2854#discussion_r1141823858. Also fixed conflicts and rebased this branch. Please recheck it one more time when you have time. And everything is OK I will merge it.
